src/lib/token.test.ts

```diff
@@ -0,0 +1,313 @@
+import {
+  Field,
+  shutdown,
+  isReady,
+  State,
+  state,
+  UInt64,
+  SmartContract,
+  Mina,
+  PrivateKey,
+  Party,
+  method,
+  PublicKey,
+  DeployArgs,
+  Permissions,
+  Token,
+  Ledger,
+} from '../../dist/server';
+
+const tokenSymbol = 'MY_TOKEN';
+
+class TokenContract extends SmartContract {
+  @state(UInt64) totalAmountInCirculation = State<UInt64>();
+  @state(UInt64) maxAmountInCirculation = State<UInt64>();
+
+  deploy(args: DeployArgs) {
+    super.deploy(args);
+    this.setPermissions({
+      ...Permissions.default(),
+      editState: Permissions.proofOrSignature(),
+    });
+  }
+
+  @method init() {
+    this.tokenSymbol.set(tokenSymbol);
+    this.totalAmountInCirculation.set(UInt64.zero);
+    this.maxAmountInCirculation.set(UInt64.from(100_000_000));
+  }
+
+  @method mint(receiverAddress: PublicKey, amount: UInt64) {
+    let totalAmountInCirculation = this.totalAmountInCirculation.get();
+    this.totalAmountInCirculation.assertEquals(totalAmountInCirculation);
+
+    let maxAmountInCirculation = this.maxAmountInCirculation.get();
+    this.maxAmountInCirculation.assertEquals(maxAmountInCirculation);
+
+    let newTotalAmountInCirculation = totalAmountInCirculation.add(amount);
+
+    newTotalAmountInCirculation.value
+      .lte(maxAmountInCirculation.value)
+      .assertTrue();
+
+    this.token().mint({
```

Comment:
Would developers need to use this assertion every time they use the mint method? 

Asking because it seems like potentially a common error case to forget this, if so. Could it be handled automatically inside of `mint()`? I wouldn't imagine anyone ever needing to deviate from this condition.

Reply:
Gregor beat me to this! All the points above are solid reasons why we shouldn't push down these custom behaviors of zkApps into the library unless there's a very good reason too. In the endgame of zkApps, I imagine there would be libraries you could import that are similar to contracts you extend in Solidity to get these checks for the developer.

Reply:
cool